src/Monad/Collection.php

```diff
@@ -45,20 +45,20 @@ public function map(callable $transformation)
      */
     public function ap(FantasyLand\Apply $applicative)
     {
-        // Sine in php List comprehension is not available, then I doing it like this
+        // Since we don't have List comprehension in PHP, use a foreach
         $result = [];
         $isCollection = $applicative instanceof Collection;
 
         foreach ($this->extract() as $value) {
-            $partial = $applicative->map($value)->extract();
+            $partial = f\valueOf($applicative->map($value));
             if ($isCollection) {
                 $result = f\push($result, $partial);
             } else {
                 $result[] = $partial;
             }
         }
 
-        return $applicative::of($result);
+        return $applicative instanceof FantasyLand\Pointed ? $applicative::of($result) : $result;
```

Comment:
But the method request an instance of `Apply` not `Applicative`.

I agree that you will probably always receive an instance of `Applicative` and the check is not really needed but I think there is no way to properly enforce this using the type hints.

I can revert the change, I mostly did it because my IDE complained.